### PR TITLE
native: *long* overdue fixes

### PR DESCRIPTION
**Pull request by LudwigKnuepfer**
Fixes for natives hwtimer and signal handling.

The hwtimer fixes made the signal handling fixes more pressing.

Some issues that should be fixed/closed by this:
- ~~https://github.com/RIOT-OS/RIOT/issues/499~~
- ~~https://github.com/RIOT-OS/RIOT/issues/715~~
- ~~https://github.com/RIOT-OS/RIOT/issues/1281~~
- ~~https://github.com/RIOT-OS/RIOT/issues/1905~~

### CODE REVIEW

**Comment by authmillenon** [comment]
@LudwigOrtmann says he needs someone to fix/comment on his ARM code. @thomaseichinger @Farthen?

**Comment by OlegHahm** [comment]
What's the state here? What's missing? And can't we already merge parts of this PR?

**Comment by LudwigOrtmann** [comment]
no change since the last comment - can't merge without breaking arm support

File: cpu/native/startup.c [hunk 1]
```diff
@@ -344,5 +344,6 @@ __attribute__((constructor)) static void startup(int argc, char **argv)
     board_init();
 
     puts("RIOT native hardware initialization complete.\n");
+    eINT();
```
Comment (gebart) [inline cpu/native/startup.c]: Can this be changed to `enableIRQ()` or are they different on native?

File: cpu/native/tramp.S [hunk 2]
```diff
@@ -28,6 +28,13 @@ __native_sig_leave_tramp:
     popfl
 
     ret
+
+.globl __native_sig_leave_handler
+__native_sig_leave_handler:
```
Comment (gebart) [inline cpu/native/tramp.S]: why double underscores on MACH and single underscores on the others?

Comment (LudwigOrtmann) [inline cpu/native/tramp.S]: Apple - think different.

Comment (OlegHahm) [inline cpu/native/tramp.S]: haha, nice one!

**Comment by authmillenon** [comment]
Those fixes are very overdue now ;-)

**Comment by OlegHahm** [comment]
@LudwigKnuepfer, could you state briefly what would be needed to fix ARM support in this PR?

**Comment by LudwigKnuepfer** [comment]
It needs to do the right thing, i.e. what x86 does.
Sorry, can't say what that is precisely off the top of my head anymore.
If there is someone with the skill-set to implement available I'm willing to help understand what this means exactly based on the x86 assembly if it's not self-evident.

**Comment by OlegHahm** [comment]
Potential candidates could be @gebart, @thomaseichinger, @kaspar030, @Yonezawa-T2.

**Comment by authmillenon** [comment]
Me, too when I'm done with my thesis.

**Comment by OlegHahm** [comment]
Same here. ;)

**Comment by thomaseichinger** [comment]
I can help with the ARM assembly as it was done by @LudwigKnuepfer and me back in the days IIRC. Don't have a clue about the rest though.

**Comment by OlegHahm** [comment]
Seems like we have a dream team.

**Comment by OlegHahm** [comment]
Seems the dream team is dreaming too much. ;-) @LudwigKnuepfer, @thomaseichinger, chances to get this done until next Wednesday are low, right?

**Comment by LudwigKnuepfer** [comment]
@OlegHahm yes, very low unless someone else works on it ;)

**Comment by emmanuelsearch** [comment]
@LudwigKnuepfer in your opinion what is bottleneck towards merging this?

**Comment by OlegHahm** [comment]
As far as I see it, we have three options:
1. Merge this as is (after fixing merge conflicts) and break ARM support for native (including Raspberry Pi) - hopefully temporarily.
2. Don't merge and live with native that is prone to crash at medium to high interrupt load (e.g. network traffic).
3. Fix the remaining issue with ARM (with consulting from @LudwigKnuepfer).

IMO 3. would be obviously the preferred solution - and 2. the least preferable.

**Comment by emmanuelsearch** [comment]
How about going for 1. now if it is a low-hanging fruit and opening an issue for 3. which would then be dealt with in a second phase?

**Comment by authmillenon** [comment]
I would prefer to go only with 3. There is no immediate need to merge this PR (1), especially since it breaks stuff.

**Comment by OlegHahm** [comment]
Well, stuff *is* broken - the question is which defect is less desirable.

**Comment by authmillenon** [comment]
True, but dropping support for a whole platform seems to be less desirable to me thaǹ a known issue we only face in *certain* situations.

**Comment by OlegHahm** [comment]
For me personally, I've never had any real use-case for running native on ARM, but I had many use-cases where this bug have become a real problem.

**Comment by authmillenon** [comment]
AFAIK the thing @Yonezawa-T2 is working on is running on Raspberry Pi.

**Comment by OlegHahm** [comment]
You could always use the latest release for that.

**Comment by kaspar030** [comment]
> I would prefer to go only with 3. There is no immediate need to merge this PR (1), especially since it breaks stuff.

+1

**Comment by OlegHahm** [comment]
So, who's gonna fix it then?

File: cpu/native/native_cpu.c [hunk 3]
```diff
@@ -187,30 +213,40 @@ void isr_thread_yield(void)
 {
     DEBUG("isr_thread_yield\n");
 
+    if (_native_sigpend > 0) {
+        DEBUG("isr_thread_yield(): handling signals\n\n");
+        native_irq_handler();
```
Comment (OlegHahm) [inline cpu/native/native_cpu.c]: This may lead to running the wrong thread. After `isr_cpu_switch_context_exit()` is called, the previous thread is restored before `sched_run()` is getting called.

**Comment by LudwigKnuepfer** [comment]
Regarding the arm-breaks - I can't really recall, but suspect it might be possible to get the x86 fixes while keeping arm in a less robust state.

**Comment by kYc0o** [comment]
any news?

**Comment by LudwigKnuepfer** [comment]
Not from my side.

**Comment by OlegHahm** [comment]
@miri64, @kaspar030, willing to reconsider? Still no one seems to be willing to work on this and it's still broken.

**Comment by OlegHahm** [comment]
Just to clarify: with this PR I was able to run an experiment with up to 1000 native instances, which I think is very desirable to make this possible in master.

**Comment by emmanuelsearch** [comment]
+1 for trying to merge that as soon as possible.
Can anyone confirm if this would completely break ARM?
Else we could indeed consider merging this which brings obvious improvements...

**Comment by OlegHahm** [comment]
If ARM support is [broken](#5644) anyway, why not just merge this?

**Comment by kaspar030** [comment]
> If ARM support is broken anyway, why not just merge this?

As far as I understand, the current ARM brokenness is a one line problem. This PR would make it harder to fix.

**Comment by kaspar030** [comment]
> Just to clarify: with this PR I was able to run an experiment with up to 1000 native instances, which I think is very desirable to make this possible in master.

Do you have a rebased version lying around?

**Comment by LudwigKnuepfer** [comment]
As this looked like a one-liner I started looking into it and there are more broken bits.

**Comment by LudwigKnuepfer** [comment]
yes, @OlegHahm has a rebased version, so do I

**Comment by OlegHahm** [comment]
> As far as I understand, the current ARM brokenness is a one line problem. This PR would make it harder to fix.

My main point was: if ARM support is broken for such a long time, it cannot be used very often. Btw would this PR **breaks** ARM support or would just the fixes not be working for ARM?

**Comment by LudwigKnuepfer** [comment]
| Btw would this PR breaks ARM support or would just the fixes not be working for ARM?

I will check this shortly, need to run ;)

**Comment by LudwigKnuepfer** [comment]
> As this looked like a one-liner I started looking into it and there are more broken bits.

OK, I found the second source of trouble:
https://github.com/RIOT-OS/RIOT/issues/5680

**Comment by LudwigKnuepfer** [comment]
Rebased on master using a recent rebase by @OlegHahm as a base ;)

**Comment by LudwigKnuepfer** [comment]
BTW: This also contains an arm-assembly fix by @thomaseichinger I did not proofread it, but at least it seems to work on some test applications and is an improvement over the plain segfault that was there prior to the change.

**Comment by OlegHahm** [comment]
No more WIP then?

**Comment by LudwigKnuepfer** [comment]
Not sure, I'm just digging out issues that might be affected by this..

**Comment by LudwigKnuepfer** [comment]
Could someone please check if any of these are fixed with this PR?
- https://github.com/RIOT-OS/RIOT/issues/3824
- https://github.com/RIOT-OS/RIOT/issues/3341
- https://github.com/RIOT-OS/RIOT/issues/2530

**Comment by LudwigKnuepfer** [comment]
> y main point was: if ARM support is broken for such a long time, it cannot be used very often. Btw would this PR breaks ARM support or would just the fixes not be working for ARM?

It would have been breaking arm support if that had not been broken already... Anyways: with the latest commit sponsored by @thomaseichinger this is not the case anymore.

**Comment by miri64** [comment]
> - https://github.com/RIOT-OS/RIOT/issues/3824

Sender does not crash after 1000000 0-delay pings (even when repeating the experiment) => I would say its fixed (though I don't have this problem in master, too :-/ so yay?... Maybe @OlegHahm can have a look)

> - #3341

Still crashes reliably.

> - #2530

Refers to `rpl_udp`... don't know how to test in post-GNRC-RIOT ;-) (but did not see this message for a while now)

**Comment by OlegHahm** [comment]
> > *        #3341
>
> Still crashes reliably.

Confirmed.

**Comment by kYc0o** [comment]
I guess if it's still crashing it won't go for this release... am I wrong?

**Comment by miri64** [comment]
AFAIK the original fixes have nothing to do with the crashing issue. It was just hope by @LudwigKnuepfer that they would be fixed also.

**Comment by miri64** [comment]
So unless this breaks something (which to my knowledge it does not) we should remove the WIP status and merge it.

**Comment by LudwigKnuepfer** [comment]
I would not say "hope" (I just dug out issues with "native" and "crash" - in retrospect at least one of them seems unrelated) but I saw a possibility that they might be affected.

**Comment by LudwigKnuepfer** [comment]
@Yonezawa-T2 could you please verify this branch on OSX? TA =)

**Comment by Yonezawa-T2** [comment]
Unfortunately, `examples/hello-world` fails with segfault.

```
* thread #1: tid = 0x12a3be, 0x00002cab hello-world.elf`sched_run + 75 at sched.c:71, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
-> 71  	    thread_t *next_thread = container_of(sched_runqueues[nextrq].next->next, thread_t, rq_entry);
```
  
`sched_runqueues[nextrq].next == NULL`
`nextrq == 0`

Rebasing on master doesn't help.

It seems that changes introduced by 1d48199e5cfff5a4148950461423cc024b8d9fd8, af24d07a06f2ff8989b8136003b19f494ceba283, and/or 3acd57b6782347b74302c17734488a44f8bd64ae breaks it.


- 74f18dec75b4197d50db67e732f42716118132fa
  fixup: integrate arm assembly patch by @eicht

  `Segmentation fault: 11`

- 76811625901dd3e08705f6486e1b580796375c10
  SM: cpu native: fix merge hickups

  `Segmentation fault: 11`

- f0575b201ed0246acdb3421d52b82323bfb45a68
  _native_sig_leave_handler arm STUB

  `cpu/native/startup.c:337:5: error: implicit declaration of function 'eINT' is invalid in C99`

  When `eINT` is replaced with `irq_enable`:
  `Segmentation fault: 11`

- 3acd57b6782347b74302c17734488a44f8bd64ae
  FIXUP: OSX/FreeBSD/arm Linux register access

  `cpu/native/startup.c:337:5: error: implicit declaration of function 'eINT' is invalid in C99`

  When `eINT` is replaced with `irq_enable`:
  `Segmentation fault: 11`

- af24d07a06f2ff8989b8136003b19f494ceba283
  FIXUP: add OSX _native_sig_leave_handler

  `cpu/native/native_cpu.c:79:41: error: member reference type 'struct __darwin_mcontext32 *' is a pointer; did you mean to use '->'?`

- 1d48199e5cfff5a4148950461423cc024b8d9fd8
  WIP: native: add syscall-leave trampoline

  `cpu/native/native_cpu.c:79:41: error: member reference type 'struct __darwin_mcontext32 *' is a pointer; did you mean to use '->'?`

- cf7114742bd8ec250bf72cdc93bcff0325694475
  native: add thread_yield_higher misuse warning

  `cpu/native/startup.c:337:5: error: implicit declaration of function 'eINT' is invalid in C99`

  When `eINT` is replaced with `irq_enable`: works fine.

**Comment by Yonezawa-T2** [comment]
Some more states:

`runqueue_bitcache: 2642527641 == 0b10011101100000011100010110011001`
```
sched_runqueues: {
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x000cd300 }, // pid == 2, name == "idle"
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x00000000 },
  { .next = 0x00091300 }, // pid == 1, name == "main"
}
```
`active_thread: NULL`
```
sched_threads: {
  0x00000000,
  0x000912f8, // pid == 1, name == "main"
  0x000cd2f8, // pid == 2, name == "idle"
  0x00000000,
  ...
}
```

**Comment by LudwigKnuepfer** [comment]
Debugging with gdb it turns out that `runqueue_bitcache` is located behind `main_thread` and violated by `getcontext` in `thread_stack_init`. I have no idea why `getcontext` does write past the memory area pointed to by `p`.

**Comment by LudwigKnuepfer** [comment]
This piece of magic fixes it:
```diff
-    p = (ucontext_t *)(stk + (stacksize - sizeof(ucontext_t)));
-    stacksize -= sizeof(ucontext_t);
+    int cs = sizeof(ucontext_t) * 4;
+    p = (ucontext_t *)(stk + (stacksize - cs));
+    stacksize -= cs;
```
Changing 4 to 2 in the diff above still results in a segfault...

**Comment by kYc0o** [comment]
Can we then move it to the next release, but let't try to make it the first merge after the release and then backport it.

**Comment by LudwigKnuepfer** [comment]
Do what you want with the release, I don't care.

**Comment by LudwigKnuepfer** [comment]
Might be fixed by rearranging includes to make sure nobody includes `ucontext.h` (transitively) before it is included in `native_internal.h` with `_XOPEN_SOURCE` exported. (Or taking some other measurement to ensure the export exists..)

**Comment by kYc0o** [comment]
Ok, actually the release question is for organisational stuff, at the the the aim is to fix it, and actually thanks to work on it, and to all who tested and helped until now.

I'll try to make more tests on OSX to see if it works.

**Comment by OlegHahm** [comment]
> Do what you want with the release, I don't care.

Reads a bit strange if maintainer writes this...

**Comment by kYc0o** [comment]
>> Do what you want with the release, I don't care.

> Reads a bit strange if maintainer writes this...

I can understand after tons of mails and a the pressure on my side...

**Comment by LudwigKnuepfer** [comment]
>>> Do what you want with the release, I don't care.

>> Reads a bit strange if maintainer writes this...

I just wanted to say that I have no opinion on whether this PR should be merged into the release. I will not be offended and don't want to postpone the date because of this PR.

> I can understand after tons of mails and the pressure on my side...

I don't feel pressured at all, you have to try harder ;-)

**Comment by LudwigKnuepfer** [comment]
> Might be fixed by rearranging includes to make sure nobody includes ucontext.h (transitively) before it is included in native_internal.h with _XOPEN_SOURCE exported. (Or taking some other measurement to ensure the export exists..)

`export CFLAGS=-D_XOPEN_SOURCE` on OSX (along with the remove of the definition in `native_cpu.c`) for building fixes the problem. The same on Linux makes the build fail. Might consider just adding it to the Makefile for OSX.

**Comment by OlegHahm** [comment]
What is the latest here?

**Comment by LudwigKnuepfer** [comment]
I'm still in favor of just ditching OSX support altogether ;) Otherwise someone should look into the solution proposal outlined in my last comment on the subject. I have no time to do it atm.

**Comment by OlegHahm** [comment]
I would also prefer to drop OSX support for native over having suboptimal (broken) native support for **all** platforms.

**Comment by Yonezawa-T2** [comment]
This works. At least, examples/hello-world, tests/gnrc_ipv6_ext, and tests/gnrc_sixlowpan run without errors.

`_DARWIN_C_SOURCE` is required for `SIGIO` constant.

```diff
diff --git a/cpu/native/Makefile.include b/cpu/native/Makefile.include
index 2ad2a35..890d558 100644
--- a/cpu/native/Makefile.include
+++ b/cpu/native/Makefile.include
@@ -6,3 +6,7 @@ ifeq ($(BUILDOSXNATIVE),1)
 endif
 
 export USEMODULE += periph
+
+ifeq ($(shell uname -s),Darwin)
+export CFLAGS += -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE
+endif
diff --git a/cpu/native/native_cpu.c b/cpu/native/native_cpu.c
index b646b4b..dc4a2ca 100644
--- a/cpu/native/native_cpu.c
+++ b/cpu/native/native_cpu.c
@@ -29,14 +29,7 @@
 #include <signal.h>
 #undef __USE_GNU
 
-
-#ifdef __MACH__
-#define _XOPEN_SOURCE
-#endif
 #include <ucontext.h>
-#ifdef __MACH__
-#undef _XOPEN_SOURCE
-#endif
 #include <err.h>
 
 #ifdef HAVE_VALGRIND_H
```

**Comment by miri64** [comment]
So this would uphold `native` support on MAC OS?

**Comment by LudwigKnuepfer** [comment]
I applied the patch by @Yonezawa-T2 above and rebased.

@miri64 I guess so, feel free to test ;)

**Comment by miri64** [comment]
> @miri64 I guess so, feel free to test ;)

Gave me the possibility to restart milhouse since the power-outage ;-)

**Comment by miri64** [comment]
(milhouse is unwilling to cooperate

```
 _TEXT  __DATA  __OBJC  others  dec     hex
 94208   1220608 0       24576   1339392 147000
 bash: objcopy: command not found
```

any hints?)

**Comment by thomaseichinger** [comment]
@miri64 I created the needed symlink, should work now.

**Comment by miri64** [comment]
`tapsetup` is not working though (maybe some tool missing too?), but I tested a few non-networking applications (examples/hello_world, examples/ipc_pingpong, examples/riot_and_cpp, examples/timer_periodic_wakeup, tests/netdev2_test, tests/netdev2_test, tests/shell) and all worked.
I used this PR on Linux several times, too. So from my perspective I would say its mergable now, though I'm a little bit hesitant because of the ARM support. What's the stance on that?

**Comment by OlegHahm** [comment]
@LudwigKnuepfer, I think we could finally move forward by squashing and merging or do you disagree?

@miri64, thanks a lot for testing.

@Yonezawa-T2, thanks for the fix!

**Comment by LudwigKnuepfer** [comment]
@OlegHahm I squashed.

**Comment by OlegHahm** [comment]
I will try to test it one last time.

**Comment by miri64** [comment]
@OlegHahm did you? If not can you summarize what you planned to do?

**Comment by OlegHahm** [comment]
Sorry. I was planning to use an application I used for some experiments with 1000 nodes. This experiments failed rather reproducable without this PR. Unfortunately, it is not based on master and I have some difficulties rebasing. I will give it another try tomorrow.

**Comment by OlegHahm** [comment]
I'm still on it.

**Comment by OlegHahm** [comment]
Unfortunately, I was not able to reproduce my experiment setup. However, I realized that flood pinging native from Linux currently result in a segfault without this PR. Merging this PR allows for floodpinging from multiple threads. Hence: ACK!

**Comment by miri64** [comment]
The long quest is at an end! :D

**Comment by LudwigKnuepfer** [comment]
Almost 2 years .. whoa.
